File: scala/collatz-conjecture/src/main/scala/CollatzConjecture.scala
import annotation.tailrec

object CollatzConjecture {

  @tailrec
  def steps(n: Int, s: Int = 0): Option[Int] = {
    if (n < 1)
      None
    else if (n == 1)
      Some(s)
    else if (n % 2 == 0)
      steps(n / 2, s + 1)
    else
      steps(3 * n + 1, s + 1)
  }
}

File: fortran/collatz-conjecture/collatz_conjecture.f90

module collatz_conjecture
  implicit none
contains

  integer function steps(n)
    integer, intent(in) :: n
    integer :: m

    if (n < 1) then
      steps = -1
      return
    end if

    m = n
    steps = 0

    do while (m > 1)
      if (mod(m, 2) == 0) then
        m = m / 2
      else
        m = m * 3 + 1
      end if
      steps = steps + 1
    end do

  end function

end module
